clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -0,0 +1,366 @@
+import com.linkedin.venice.schema.SchemaData;
+import com.linkedin.venice.schema.SchemaEntry;
+import com.linkedin.venice.serializer.RecordDeserializer;
+import com.linkedin.venice.serializer.SerializerDeserializerFactory;
+import com.linkedin.venice.utils.PartitionUtils;
+import com.linkedin.venice.utils.Utils;
+import com.linkedin.venice.utils.VeniceProperties;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.util.Utf8;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class RequestBasedMetadata extends AbstractStoreMetadata {
+  private static final Logger LOGGER = LogManager.getLogger(RequestBasedMetadata.class);
+  private static final String VERSION_PARTITION_SEPARATOR = "_";
+  private static final long ZSTD_DICT_FETCH_TIMEOUT = 10;
+  private static final long DEFAULT_REFRESH_INTERVAL_IN_SECONDS = 60;
+  private static final long INITIAL_UPDATE_CACHE_TIMEOUT_IN_SECONDS = 30;
+  private static final long RETRY_WAIT_TIME_IN_MS = 1000;
+  private final long refreshIntervalInSeconds;
+  private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+  private final AtomicInteger currentVersion = new AtomicInteger();
+  private final AtomicInteger latestSuperSetValueSchemaId = new AtomicInteger();
+  private final AtomicReference<SchemaData> schemas = new AtomicReference<>();
+  private final Map<String, List<String>> readyToServeInstancesMap = new VeniceConcurrentHashMap<>();
+  private final Map<Integer, PartitionerPair> versionPartitionerMap = new VeniceConcurrentHashMap<>();
+  private final Map<Integer, ByteBuffer> versionZstdDictionaryMap = new VeniceConcurrentHashMap<>();
+  private final CompressorFactory compressorFactory;
+  private final D2TransportClient transportClient;
+  private final D2ServiceDiscovery d2ServiceDiscovery;
+  private final String routerD2ServiceName;
+  private final ClusterStats clusterStats;
+  private volatile boolean isServiceDiscovered;
+
+  public RequestBasedMetadata(
+      ClientConfig clientConfig,
+      D2TransportClient transportClient,
+      D2ServiceDiscovery d2ServiceDiscovery,
```

Comment:
`D2ServiceDiscovery` was only added so I can mock the behavior of `D2ServiceDiscovery.find()` in tests. I'm only ever passing in the object with the default constructor. Is it okay to have this param just for that or is there another approach?

Reply:
If it's test only I think it's better to expose it as a setter method to override the internal `D2ServiceDiscovery` instead of adding to to the constructor.